### DISCUSSION TITLE
NewRealm.All returns collection with nullable ID-s in the migration function

### DISCUSSION THREAD

**Discussion by vardansargsyan92** [realm/realm-dotnet#2521] [Q&A] [answered]
Hello. 

Recently I have modified my project's entities to be able to support realm sync. I have added PartitionKey in the all realm objects and have added a new migration function (Scheme version let's say 4) supporting that functionality. Also I have many migration functions for different already refactored cases. Now I am running app on old instance of database and when the previous migration functions start (Let's say functions for schema versions 1, 2, 3) then the newRealm.All<...>() method returns realm objects with nullable ID-s. 

So the question is what is wrong there?. Did I missing something?

p.s. I am using the latest versions of Realm, Xamarin, Fody Nuget packages.

**Comment by papafe** [comment, marked as answer]
Hi @vardansargsyan92, would it be possible for you to post the migration function to see exactly what you're doing? It's a little bit difficult to judge what is happening otherwise.

As a separate note, I don't know if you're aware of it, but when you decide to use sync unfortunately you can't simply use a migration, but would need to copy the objects manually to the sync realm. That happens because local realms and synced ones are not compatible with each other.

**Reply by vardansargsyan92** [reply]
Hello @papafe. 
 
Thank you for your instant response. Yes, you are right. We can't simply use migration and, we should copy objects manually to the realm sync. The process which I am doing now is a kind of intermediate process to prepare realm objects to sync functionality (Adding partition keys, because sync works with partition keys). Actually I have already found the reason.

In my case the reason of getting nullable ids is the following:

Imagine that we have an app with a local database with already written two migration functions. The current database's schema version is 3. Now we decide to prepare the realm objects to realm-sync. So we add partition keys to realm objects. 

```
        [PrimaryKey]
        [MapTo("_id")]
        public string Id { get; set; } = Guid.NewGuid().ToString();

        [Required]
        [MapTo("_partitionKey")]
        public string PartitionKey { get; set; }
```

Now, we increase the database scheme to 4 and write a migration function supporting the latest changes. Now, when we run the app and migration starts according to these changes the new realm objects become with nullable id's because the newly added migration function is called last. So first, we should call the last migration function and the remaining functions.

**Reply by papafe** [reply]
I am still a little bit unsure about it, but I am glad that you found a solution to your issue!
